Added runnableExamples and sample code for algorithm

ThomasTJdev opened the pull request:
Updated documentation for algorithm module:
- Code sample
- See also links
- Runnable examples

**Notes**
1) I could not find a way to use `runnableExamples` with `template sortedByIt` - it got messy.
2) I would have liked to add some more intro text, but I couldn't come up with any useful.

refs #10330
CC @narimiran

narimiran commented:
The procs which have `cmp: proc(...)` in their argument list allow you to pass your custom proc which is used to compare two elements, making it possible to have "custom sorting".
For example, if you have a seq of tuples or a seq of objects, you can then sort it by some particular field.

This should be visible from the examples for those procs (so, some custom sorting proc needs to be defined).


----

> I started to add see also links for "identical" procs, but looking at your changes in strutils, you are not making links between the different `split()`. I have removed these links from my commit.

Feel free to include those. (I did it for [strutils.toLowerAscii](https://nim-lang.github.io/Nim/strutils.html#toLowerAscii%2Cchar) and some others)

----

> This question is related with 3). I'm in doubt about the link text. E.g.:
> ```## * `lowerBound proc<#lowerBound,openArray[T],K,proc(T,K)>`_ sorted by ``cmp`` in the specified order```

In cases like this, I would write something like:

```
## * `lowerBound proc<#lowerBound,openArray[T],K>`_ with the default (`system.cmp`) sorting
## * `lowerBound proc<#lowerBound,openArray[T],K,proc(T,K)>`_ with custom sorting
```

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -286,6 +352,20 @@ func sort*[T](a: var openArray[T],
+  ## * `sort proc<#sort,openArray[T]>`_
+  ## * `sorted proc<#sorted,openArray[T],proc(T,T)>`_ sorted by ``cmp`` in the specified order
+  ## * `sorted proc<#sorted,openArray[T]>`_
+  ## * `sortedByIt template<#sorted,openArray[T]>`_
```
wrong link ;)

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -272,7 +339,6 @@ func sort*[T](a: var openArray[T],
   ## .. code-block:: nim
```
I cannot comment above (line 332):

This need some breathing room by separating paragraphs. Add an empty comment line like this:

```
  ## is guaranteed to be stable and the worst case is guaranteed to
  ## be O(n log n).
  ##
  ## The current implementation uses an iterative
  ## mergesort to achieve this. It uses a temporary sequence of
```

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -31,28 +52,32 @@ template fillImpl[T](a: var openArray[T], first, last: int, value: T) =
 
 proc fill*[T](a: var openArray[T], first, last: Natural, value: T) =
-  ## fills the slice ``a[first..last]`` with ``value``.
+  ## Fills the slice ``a[first..last]`` with ``value``.
```
What happens if I do:
```nim
import algorithm

var a: array[5, int]
a.fill(3, 10, 9)
```

One should not need to guess/try stuff like that.

If this works, then it should be shown in the example (see for example [math.log](https://nim-lang.github.io/Nim/math.html#log%2CT%2CT), where I was curious to see what happens if I give `x<=0` to `log`. 
If it raises an error, it should be mentioned in the docs (Something along the lines: "If an invalid range is passed, it raises `IndexError`"), and you can then write an example like this:

```nim
var a: array[5, int]
doAssertRaises(IndexError, a.fill(3, 10, 9))
```

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -401,15 +540,16 @@ proc product*[T](x: openArray[seq[T]]): seq[seq[T]] =
-  ##     echo v # @[0, 1, 2, 3, 4, 5, 6, 7, 9, 8]
+  ## **See also:**
+  ## * `prevPermutation proc<#prevPermutation,openArray[T]>`_
+  runnableExamples:
```
I'd rather see an example of a smaller array/seq permuted several times.

What is also interesting about this proc is that if you start with unsorted array/seq, the repeated permutations will **not** give you all permutations, but they will stop with the "last" one. (try `nextPermutation` on `[3, 2, 1]`)

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -561,36 +702,54 @@ proc rotateLeft*[T](arg: var openarray[T]; slice: HSlice[int, int]; dist: int):
-      var a = [1, 2, 3, 4, 5]
-      a.rotateLeft(2)
-      doAssert a == [3, 4, 5, 1, 2]
+    var a = [1, 2, 3, 4, 5]
```
As above, an example with multiple rotations (showing the result after each rotation) would be more interesting/informative.

narimiran replied:
Also, in one of these rotating procs, show an example where you use negative `dist`.

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -8,19 +8,40 @@
+##    echo "John was born as number " & $(john + 1) # --> 2
+##
+##
+## See also:
```
If there are no other headings before it (e.g. "Basic usage"), then I didn't use heading for this, just `**See also:**`. (And when I turn this into a heading, then I delete the `:`, heading looks better without it, IMO)

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -8,19 +8,40 @@
+## .. code-block::
+##    import algorithm
+##
+##    var birth = ["2000 John", "1995 Marie", "2010 Jane"]
```
If you turn this into `seq[tuple[year: int, name: string]]` it might be easier to showcase the difference between `sort` with the default `system.cmp` and a `sort` where you defined your own `cmp`, something like this:

```nim
import algorithm

type People = tuple
  year: int
  name: string

var a: seq[People]

a.add((2000, "John"))
a.add((2005, "Marie"))
a.add((2010, "Jane"))

a.sort()
assert a == @[(year: 2000, name: "John"), (year: 2005, name: "Marie"), (year: 2010, name: "Jane")]

proc myCmp(x, y: People): int =
  if x.name < y.name: -1 else: 1

a.sort(myCmp)
assert a == @[(year: 2010, name: "Jane"), (year: 2000, name: "John"), (year: 2005, name: "Marie")]
```

(You might want to reuse some of these ideas for later, when you describe those functions allowing for a custom `cmp`)

ThomasTJdev commented:
**Question:** when there's 2 kind of examples with many lines, it can get a little messy to look at in the final documentation. Is that just how life is, or should we split it into 2 runnableExamples?

```nim
  runnableExamples:
    var
      a = [1, 3, 4, 2, 5]
      b = ["boo", "foo", "bar", "qux"]
      c = ["boo", "foo", "bar", "qux"]
    sort(a, system.cmp[int])
    assert a == [1, 2, 3, 4, 5]
    sort(b, system.cmp)
    assert b == ["bar", "boo", "foo", "qux"]
  runnableExamples:
    var c = ["boo", "foo", "bar", "qux"]
    proc myCmp(x, y: string): int =
      if x == "bar": 1 else: -1
    sort(c, myCmp)
    assert c == ["boo", "foo", "qux", "bar"]
```

narimiran commented:
> Is that just how life is, or should we split it into 2 runnableExamples?

Hmmm, I haven't had that problem yet (or if I had, I didn't noticed and just used one `runnableExamples`).

Based on the above screenshots, you're trying to show the example of `sort` where you can pass a custom `cmp` — you can leave out the first part, where you pass `system.cmp` (that's what is used in `sort` with the default `cmp` param), and just show the example with your custom `cmp` proc (which doesn't compare the elements, just checks if the first element in the comparison is `"bar"`!?).

ThomasTJdev commented:
> > Is that just how life is, or should we split it into 2 runnableExamples?
> 
> Hmmm, I haven't had that problem yet (or if I had, I didn't noticed and just used one `runnableExamples`).
> 
> Based on the above screenshots, you're trying to show the example of `sort` where you can pass a custom `cmp` — you can leave out the first part, where you pass `system.cmp` (that's what is used in `sort` with the default `cmp` param), and just show the example with your custom `cmp` proc (which doesn't compare the elements, just checks if the first element in the comparison is `"bar"`!?).

You right. Removed first examples and updated `myCmp` to compare the values.

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -208,7 +242,10 @@ proc lowerBound*[T, K](a: openArray[T], key: K, cmp: proc(x: T, k: K): int {.clo
   ## * `upperBound proc<#upperBound,openArray[T],K,>`_
```
Wrong link.

It should be:
```
   ## * `upperBound proc<#upperBound,openArray[T],T>`_
```

The same goes for a link to `lowerBound` on the line 289.

ThomasTJdev replied:
56c897f968f9215006521e10f310bd03d09901fe

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -357,15 +400,14 @@ func sort*[T](a: var openArray[T],
   ## * `sorted proc<#sorted,openArray[T],proc(T,T)>`_ sorted by ``cmp`` in the specified order
   ## * `sorted proc<#sorted,openArray[T]>`_
-  ## * `sortedByIt template<#sorted,openArray[T]>`_
+  ## * `sortedByIt template<#sortedByIt.t,untyped,untypedd,openArray[T]>`_
```
Wrong link. It should be:

```
#sortedByIt.t,untyped,untyped
```

ThomasTJdev replied:
Ohh :-/ a716c94e0efa61088c094e7ef52c3f84de8aebfe

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -732,11 +802,23 @@ proc rotatedLeft*[T](arg: openarray[T]; slice: HSlice[int, int], dist: int): seq
+  ##
   ## **See also:**
   ## * `rotateLeft proc<#rotateLeft,openArray[T],HSlice[int,int],int>`_
   ## * `rotateLeft proc<#rotateLeft,openArray[T],int>`_
```
Since there are two procs of the same name, and two of the similar name, I would add an explanation what each does to reduce confusion so the reader immediately knows which is which without the need to click first:

```
  ## **See also:**
  ## * `rotateLeft proc<#rotateLeft,openArray[T],HSlice[int,int],int>`_ for the in-place version of this proc
  ## * `rotatedLeft proc<#rotatedLeft,openArray[T],int>`_ for a version which rotates the whole container
```

(Notice that the second link is changed to `rotatedLeft`)

ThomasTJdev replied:
Updated link description: 20067e844058b0bb96b415b5a1fe8b27c94019db

narimiran commented on lib/pure/algorithm.nim:
```diff
@@ -9,20 +9,36 @@
-## =========
+## See also
+## ========
 ## * `tables module<tables.html>`_ for sorting tables
```
I think that link to `sequtils` might be of interest too.

ThomasTJdev replied:
Ok? acd0c5635934bcf5d5bf0f3231abd26e7fab5a66
